### ISSUE TITLE
Cloud Blob Request Option setConcurrentRequestCount is not working properly

### ISSUE THREAD

**Issue by chhajerpawan** [Azure/azure-storage-java#462]
### Which service(blob, file, queue, table) does this issue concern?
Storage Blob

### Which version of the SDK was used?

5.0.0
### What problem was encountered?
Its not reflecting any performance 

### Have you found a mitigation/solution?
No

**Comment by chhajerpawan** [comment]
I am trying to download a large file around 4 GB of file from Azure Blob Storage to On premises Machine 

I tried below configuration 

CloudBlockBlob blob = container.getBlockBlobReference("test1.csv");
			BlobRequestOptions option = new BlobRequestOptions();
			option.setConcurrentRequestCount(20);

But looks like file download is taking same time without this option and with this option

**Comment by rickle-msft** [comment]
@chhajerpawan Thank you for posting this issue. Just as a sanity check, I assume you are successfully passing these options into the call that retrieves the blob. May I also ask why you are using v5.0.0. Could you try upgrading and see if this problem persists?

**Comment by chhajerpawan** [comment]
I have already implemented application, If i upgrade to v11 then i need to do lot of code change 

I have pass this option to

**Comment by chhajerpawan** [comment]
blobClient.setDefaultRequestOptions(option);

**Comment by chhajerpawan** [comment]
storageAccount = CloudStorageAccount.parse(storageConnectionString);
			blobClient = storageAccount.createCloudBlobClient();
			blobClient.getDefaultRequestOptions().setConcurrentRequestCount(20);
			System.out.println(blobClient.getDefaultRequestOptions().getConcurrentRequestCount());
			container = blobClient.getContainerReference("container-for-dev");

			// Create the container if it does not exist with public access.
			System.out.println("Creating container: " + container.getName());
			//container.createIfNotExists(BlobContainerPublicAccessType.CONTAINER, new BlobRequestOptions(), new OperationContext());		    

			//Creating a sample file
			sourceFile = File.createTempFile("test1", ".csv");
			System.out.println("Creating a sample file at: " + sourceFile.toString());
			/*Writer output = new BufferedWriter(new FileWriter(sourceFile));
			 * 
			output.write("Hello Azure!");
			output.close();*/

			//Getting a blob reference
			CloudBlockBlob blob = container.getBlockBlobReference("test1.csv");
			BlobRequestOptions option = new BlobRequestOptions();
			option.setConcurrentRequestCount(20);
			//option.setSingleBlobPutThresholdInBytes(1024*1024);
			blobClient.setDefaultRequestOptions(option);
			//Creating blob and uploading file to it
			System.out.println("Uploading the sample file ");
			//blob.uploadFromFile(sourceFile.getAbsolutePath());

			//Listing contents of container
			/*for (ListBlobItem blobItem : container.listBlobs()) {
			System.out.println("URI of blob is: " + blobItem.getUri());
		}*/

		// Download blob. In most cases, you would have to retrieve the reference
		// to cloudBlockBlob here. However, we created that reference earlier, and 
		// haven't changed the blob we're interested in, so we can reuse it. 
		// Here we are creating a new file to download to. Alternatively you can also pass in the path as a string into downloadToFile method: blob.downloadToFile("/path/to/new/file").
		long v1 = new Date().getTime();
			System.out.print(new Date());
			downloadedFile = new File(sourceFile.getParentFile(), "downloadedFile.txt");
			blob.setStreamMinimumReadSizeInBytes(256*1024);
		blob.downloadToFile(downloadedFile.getAbsolutePath(),null,option,null);
		System.out.println(new Date().getTime()-v1);

**Comment by rickle-msft** [comment]
Sorry, when I said upgrade, I meant upgrade to the latest version of the sync sdk, which is version 8.1.0

**Comment by rickle-msft** [comment]
I will work on reproing this in the next couple days and get back to you

**Comment by chhajerpawan** [comment]
I did that, but seems like concurrent request is not working

**Comment by chhajerpawan** [comment]
For 500 MB file  using Azcopy I am able to download in 08 second , but if try to download using java sdk 8.1.0 sdk code based then It is taking 2 minutes to download same file.

**Comment by chhajerpawan** [comment]
Any update on this ?

**Comment by rickle-msft** [comment]
@chhajerpawan Sorry for the delay. Setting the streamMinimumReadSize to 256k looks a little suspicious as that means there is a high likelihood you are only downloading 256k on each network request. Could you try increasing this value and see if you get better performance (the limit is 4mb).

**Comment by chhajerpawan** [comment]
I tried that also but no improvement

**Comment by chhajerpawan** [comment]
setConcurrentRequestCount is not working properly , because if i change no of thread there is no difference if i change it to 1 or 10 same result for big file

**Comment by rickle-msft** [comment]
I don't believe set concurrentRequestCount is actually used in the download path at all, so it makes sense that you are not seeing a difference based on this value. That is why I am trying to investigate other factors which could be contributing to the slow download speeds you are seeing.

**Comment by chhajerpawan** [comment]
WHAT IS PURPOSE OF concurrentRequestCount ? 

If file size is big then azure will automatically chunk those file based on max chunk size and parallel download based on concurrentRequestCount  threads

**Comment by rickle-msft** [comment]
concurrentRequestCount is used in upload scenarios. Download paths do not make use of parallelization from what I am seeing. I can point you to the relevant files in the source if you would like to take a look to confirm. I am also happy to continue to look at other factors that may be playing into the performance issues you are seeing.
